substrate/frame/dap/src/lib.rs

```diff
@@ -0,0 +1,469 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Dynamic Allocation Pool (DAP) Pallet
+//!
+//! This pallet implements `FundingSink` to collect funds into a buffer account instead of burning
+//! them. The buffer account is created via `inc_providers` at genesis or on runtime upgrade,
+//! ensuring it can receive any amount including those below ED.
+//!
+//! Future phases will add:
+//! - `FundingSource` (request_funds) for pulling funds
+//! - Issuance curve and minting logic
+//! - Distribution rules and scheduling
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{
+		fungible::{Balanced, Credit, Inspect, Mutate},
+		tokens::{Fortitude, FundingSink, Precision, Preservation},
+		Currency, Imbalance, OnUnbalanced,
+	},
+	PalletId,
+};
+
+pub use pallet::*;
+
+const LOG_TARGET: &str = "runtime::dap";
+
+/// Type alias for balance.
+pub type BalanceOf<T> =
+	<<T as Config>::Currency as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::sp_runtime::traits::AccountIdConversion;
+
+	/// The in-code storage version.
+	const STORAGE_VERSION: frame_support::traits::StorageVersion =
+		frame_support::traits::StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The currency type.
+		type Currency: Inspect<Self::AccountId>
+			+ Mutate<Self::AccountId>
+			+ Balanced<Self::AccountId>;
+
+		/// The pallet ID used to derive the buffer account.
+		///
+		/// Each runtime should configure a unique ID to avoid collisions if multiple
+		/// DAP instances are used.
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Get the DAP buffer account
+		/// NOTE: We may need more accounts in the future, for instance, to manage the strategic
+		/// reserve. We will add them as necessary, generating them with additional seed.
+		pub fn buffer_account() -> T::AccountId {
+			T::PalletId::get().into_account_truncating()
+		}
+
+		/// Ensure the buffer account exists by incrementing its provider count.
+		///
+		/// This is called at genesis and on runtime upgrade.
+		/// It's idempotent - calling it multiple times is safe.
+		pub fn ensure_buffer_account_exists() {
+			let buffer = Self::buffer_account();
+			if !frame_system::Pallet::<T>::account_exists(&buffer) {
+				frame_system::Pallet::<T>::inc_providers(&buffer);
+				log::info!(
+					target: LOG_TARGET,
+					"Created DAP buffer account: {buffer:?}"
+				);
+			}
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<frame_system::pallet_prelude::BlockNumberFor<T>> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
```

Comment:
We generally don't implement this on `Pallet` but on a separate struct that we pass via runtime to the executive.
See https://github.com/paritytech/polkadot-sdk/blob/master/cumulus/parachains/runtimes/assets/asset-hub-westend/src/lib.rs#L1467

Reply:
Without any versioning, this will run on every runtime ugrade, which honestly is fine for this pallet. It will noop from the second time and runtime upgrades are not frequent. Was that your intention?